contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+        for line in self.filtered('contract_id.recurring_invoices'):
+            if not line.date_start:
+                raise ValidationError(
+                    _("You must supply a start date for contract '%s'")
+                    % line.contract_id.name
+                )
+
+    @api.constrains('date_start', 'date_end')
+    def _check_start_end_dates(self):
+        for line in self.filtered('date_end'):
+            if line.date_start and line.date_end:
+                if line.date_start > line.date_end:
+                    raise ValidationError(
+                        _(
+                            "Contract '%s' start date can't be later than "
+                            "end date"
+                        )
+                        % line.contract_id.name
+                    )
+
+    @api.depends('recurring_next_date', 'date_start', 'date_end')
+    def _compute_create_invoice_visibility(self):
+        today = fields.Date.today()
+        for line in self:
+            if line.date_start:
+                if today < line.date_start:
+                    line.create_invoice_visibility = False
+                elif not line.date_end:
+                    line.create_invoice_visibility = True
+                elif line.recurring_next_date:
+                    if line.recurring_invoicing_type == 'pre-paid':
+                        line.create_invoice_visibility = (
+                            line.recurring_next_date <= line.date_end
+                        )
+                    else:
+                        line.create_invoice_visibility = (
+                            line.recurring_next_date
+                            - line.get_relative_delta(
+                                line.recurring_rule_type,
+                                line.recurring_interval,
+                            )
+                        ) <= line.date_end
+
+    @api.model
+    def recurring_create_invoice(self, contract=False):
+        domain = []
+        date_ref = fields.Date.today()
+        if contract:
+            contract.ensure_one()
+            date_ref = contract.recurring_next_date
+            domain.append(('contract_id', '=', contract.id))
+
+        domain.extend(
+            [
+                ('contract_id.recurring_invoices', '=', True),
+                ('recurring_next_date', '<=', date_ref),
+                ('is_canceled', '=', False),
+                # '|',
+                # ('date_end', '=', False),
+                # ('date_end', '>=', date_ref),
+                # with this leaf, it's impossible to invoice the last period
+                # in post-paid case.
+                # i.e: date_end = 15/03 recurring_next_date = 31/03
+                # A solution for this, is to only check recurring_next_date
+                # and filter with create_invoice_visibility
+            ]
+        )
+        lines = self.search(domain).filtered('create_invoice_visibility')
```

Comment:
If I understand well, this is going to return all contract lines with recurring_next_date in the past, and then filter on create_invoice_visibility, so this will get slower and slower as past lines accumulate.

Would it not be much simpler to set recurring_next_date to False when there is nothing more to invoice?